src/python/grpcio/grpc/experimental/aio/_metadata.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of the metadata abstraction for gRPC Asyncio Python."""
+from typing import List, Tuple, AnyStr, Iterator, Any
+from collections import abc, OrderedDict
+
+
+class Metadata(abc.Mapping):
+    """Metadata abstraction for the asynchronous calls and interceptors.
+
+    The metadata is a mapping from str -> List[str]
+
+    Traits
+        * Multiple entries are allowed for the same key
+        * The order of the values by key is preserved
+        * Getting by an element by key, retrieves the first mapped value
+        * Supports an immutable view of the data
+        * Allows partial mutation on the data without recreating the new object from scratch.
+    """
+
+    def __init__(self, *args: Tuple[str, AnyStr]) -> None:
+        self._metadata = OrderedDict()
```

Comment:
``OrderedDict`` is guaranteed to work regardless of the version being used. Wouldn't we prefer that (because it's safer and simpler)?

Reply:
Nothing urgent since its hidden to the user, we can explore customizing and maybe improving the data-structure used under the hood, later on. So not a stopper for me!